equinox/nn/_embedding.py

```diff
@@ -89,3 +93,253 @@ def __call__(
+        key: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ):
+        """**Arguments:**
+
+        `RotaryPositionalEmbedding` requires:
+
+        - `embedding_size`: Size of the token embeddings. Must be non-negative.
+        - `max_seq_len`: The maximum sequence length. Must be non-negative if provided.
+        - `key`: Not used; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+        """
+        super().__init__(**kwargs)
+        if embedding_size < 0:
+            raise ValueError("embedding_size must not be negative.")
+        if max_seq_len is not None and max_seq_len < 0:
+            raise ValueError("max_seq_len must not be negative.")
+        self.embedding_size = embedding_size
+        self.max_seq_len = max_seq_len
+
+        if self.max_seq_len is not None:
+            if (self.embedding_size, self.max_seq_len) in internal_rope_embedding_cache:
+                self.freqs_cis = internal_rope_embedding_cache[
+                    (self.embedding_size, self.max_seq_len)
+                ]
+            else:
+                self.freqs_cis = self.precompute_freqs_cis(
+                    embedding_size, self.max_seq_len
+                )
+                internal_rope_embedding_cache[
+                    (self.embedding_size, self.max_seq_len)
+                ] = self.freqs_cis
+        else:
+            self.freqs_cis = None
+
+    @staticmethod
+    def negate_half(x: Float[Array, "max_seq_len embedding_size"]):
+        d_2 = x.shape[-1] // 2
+        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)
+
+    @staticmethod
+    def precompute_freqs_cis(
+        embedding_size: int, end: int, theta: float = 10000.0
+    ) -> Complex[Array, "end embedding_size/2"]:
+        def polar(abs, angle):
+            return jnp.array(
+                abs * jnp.cos(angle) + abs * jnp.sin(angle) * 1j, dtype=jnp.complex64
+            )
+
+        freqs = 1.0 / (
+            theta ** (jnp.arange(0, embedding_size, 2)[jnp.newaxis, :] / embedding_size)
+        )
+        t = jnp.arange(end)
+        freqs_outer = jnp.outer(t, freqs)
+        freqs_cis = polar(jnp.ones_like(freqs_outer), freqs_outer)
+        return freqs_cis
+
+    @jax.named_scope("eqx.nn.RotaryPositionalEmbedding")
+    def __call__(
+        self,
+        x: Float[Array, "seq_len embedding_size"],
+        *,
+        key: Optional[PRNGKeyArray] = None,
+    ) -> Float[Array, "max_seq_len embedding_size"]:
+        """**Arguments:**
+
+        - `x`: A JAX array of shape `(seq_len, embedding_size)`.
+        - `key`: Ignored; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+
+        **Returns:**
+
+        A JAX array of shape `(seq_len, embedding_size)`, with the rotary positional
+        encoding applied to the input.
+        """
+
+        seq_len, embedding_size = x.shape
+        assert embedding_size == self.embedding_size, (
```

Comment:
These should be `ValueError`s rather than `assert` statements. The former is used when a user has gotten something wrong; the latter is used when the developer has gotten something wrong (and an internal invariant no longer holds).